x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/Authenticator.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authc;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.settings.SecureString;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationToken;
+import org.elasticsearch.xpack.core.security.authc.Realm;
+import org.elasticsearch.xpack.core.security.user.User;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public interface Authenticator {
+
+    String name();
+
+    AuthenticationToken extractCredentials(Context context);
+
+    /**
+     * Whether authentication can proceed to authenticate as anonymous or fallback user
+     * if this authenticator can extract a credentials.
+     */
+    default boolean canFollowedByNullTokenHandler() {
+        return true;
+    }
+
+    void authenticate(Context context, ActionListener<Result> listener);
+
+    /**
+     * Gets the token from the <code>Authorization</code> header if the header begins with
+     * <code>Bearer </code>
+     */
+    static SecureString extractBearerTokenFromHeader(ThreadContext threadContext) {
+        String header = threadContext.getHeader("Authorization");
+        if (Strings.hasText(header) && header.regionMatches(true,
+            0,
+            "Bearer ",
+            0,
+            "Bearer ".length()) && header.length() > "Bearer ".length()) {
+            char[] chars = new char[header.length() - "Bearer ".length()];
+            header.getChars("Bearer ".length(), header.length(), chars, 0);
+            return new SecureString(chars);
+        }
+        return null;
+    }
+
+    class Context implements Closeable {
+        private final ThreadContext threadContext;
+        private final AuthenticationService.AuditableRequest request;
+        private final User fallbackUser;
+        private final boolean allowAnonymous;
+        private final Realms realms;
+        private final List<AuthenticationToken> authenticationTokens = new ArrayList<>();
+        private final List<String> unsuccessfulMessages = new ArrayList<>();
+        private boolean handleNullToken = true;
+        private SecureString bearerString = null;
+        private List<Realm> defaultOrderedRealmList = null;
+        private List<Realm> unlicensedRealms = null;
```

Comment:
Agreed. Many were added out of fear for performance. For example, The `bearerString` is to avoid extracting it multiple times. There is no proof for performance issue if it does get extracted 2 or 3 times. But I was more comfortable to maintain the same execution logic as much as possible. We can definitely tweak these later.